File: spring-practice/src/main/java/com/example/springpractice/product/ProductService.java
package com.example.springpractice.product;

import com.example.springpractice.common.exception.NotFoundException;
import com.example.springpractice.product.dto.ProductCreateRequest;
import com.example.springpractice.product.dto.ProductResponse;
import com.example.springpractice.product.dto.ProductUpdateRequest;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ProductService {

    private final ProductRepository productRepository;

    @Transactional
    public ProductResponse create(ProductCreateRequest request) {
        Product saved = productRepository.save(request.toEntity());
        return ProductResponse.from(saved);
    }

    public List<ProductResponse> findAll() {
        return productRepository.findAll().stream()
                .map(ProductResponse::from)
                .toList();
    }

    public ProductResponse findById(Long id) {
        return ProductResponse.from(getProductOrThrow(id));
    }

    @Transactional
    public ProductResponse update(Long id, ProductUpdateRequest request) {
        Product product = getProductOrThrow(id);
        // 더티 체킹: 트랜잭션 종료 시점에 변경 감지로 자동 UPDATE
        product.update(request.name(), request.price(), request.stockQuantity());
        return ProductResponse.from(product);
    }

    @Transactional
    public void delete(Long id) {
        Product product = getProductOrThrow(id);
        productRepository.delete(product);
    }

    private Product getProductOrThrow(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("상품을 찾을 수 없습니다. id=" + id));
    }
}

File: kotlin-practice/src/main/kotlin/com/example/kotlinpractice/product/ProductService.kt
package com.example.kotlinpractice.product

import com.example.kotlinpractice.common.exception.NotFoundException
import com.example.kotlinpractice.product.dto.ProductCreateRequest
import com.example.kotlinpractice.product.dto.ProductResponse
import com.example.kotlinpractice.product.dto.ProductUpdateRequest
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class ProductService(
    private val productRepository: ProductRepository
) {

    @Transactional
    fun create(request: ProductCreateRequest): ProductResponse {
        val saved = productRepository.save(request.toEntity())
        return ProductResponse.from(saved)
    }

    fun findAll(): List<ProductResponse> =
        productRepository.findAll().map { ProductResponse.from(it) }

    fun findById(id: Long): ProductResponse =
        ProductResponse.from(getProductOrThrow(id))

    @Transactional
    fun update(id: Long, request: ProductUpdateRequest): ProductResponse {
        val product = getProductOrThrow(id)
        // 더티 체킹: 트랜잭션 종료 시점에 변경 감지로 자동 UPDATE
        product.update(request.name, request.price, request.stockQuantity)
        return ProductResponse.from(product)
    }

    @Transactional
    fun delete(id: Long) {
        val product = getProductOrThrow(id)
        productRepository.delete(product)
    }

    private fun getProductOrThrow(id: Long): Product =
        productRepository.findById(id)
            .orElseThrow { NotFoundException("상품을 찾을 수 없습니다. id=$id") }
}
